src/components/cleaner.ts

```diff
@@ -44,23 +45,42 @@ export class Cleaner {
         if (configuration.get('latex.clean.subfolder.enabled') as boolean) {
             globs = globs.map(globType => './**/' + globType)
         }
+        globs = Array.from(new Set(globs))  // remove duplicate entries
         this.extension.logger.addLogMessage(`Clean glob matched files: ${JSON.stringify({globs, outdir})}`)
-        const files = globs.map(g => glob.sync(g, {cwd: outdir}))
-            // Reduce the array of arrays to a single array containing all the files that should be deleted
-            .reduce((all, curr) => all.concat(curr), [])
-            // Resolve the absolute filepath for every file
-            .map(file => path.resolve(outdir, file))
-        for (const file of files) {
+        // All glob patterns for folders, the glob should be end with a slash (`path.sep`) and the last component should not contain `**`
+        const folderGlobs: string[] = globs.filter(globType => globType.endsWith(path.sep) && !path.basename(globType).includes('**'))
+        const matchFolderGlobs = (folder: string): boolean => folderGlobs.some(globType => minimatch(folder, globType))
+        const pathSet: Set<string> = new Set(  // unique paths
+            globs.map(g => glob.sync(g, {cwd: outdir}))
+                 // Reduce the array of arrays to a single array containing all the files that should be deleted
+                 .reduce((all, curr) => all.concat(curr), [])
+        )
+        const paths: string[] = Array.from(pathSet)
+                                     // Sort by descending dictionary order, make sure the folder is sorted after the files in it
+                                     .sort((a, b) => b.localeCompare(a))
+                                     // Pattern `abc/**` matches folder "abc" while pattern `abc/` matches "abc/"
+                                     // Remove the duplicates and keep the folder paths with tailing slashes
+                                     .filter((p) => !pathSet.has(p + path.sep))
+        // Resolve absolute path before removing anything (for symlinked paths)
+        const entries = paths.map((relativePath) => [path.resolve(outdir, relativePath) , relativePath])
+        for (const [realPath, relativePath] of entries) {
             try {
-                const stats: fs.Stats = fs.statSync(file)
+                const stats: fs.Stats = fs.statSync(realPath)
                 if (stats.isFile()) {
-                    await fs.promises.unlink(file)
-                    this.extension.logger.addLogMessage(`Cleaning file: ${file}`)
+                    await fs.promises.unlink(realPath)
+                    this.extension.logger.addLogMessage(`Cleaning file: ${realPath}`)
+                } else if (stats.isDirectory() && matchFolderGlobs(relativePath)) {
```

Comment:
> All glob things shall be before the for loop.

The paths is globed at line 53:

```javascript
const pathSet: Set<string> = new Set(  // unique paths
    globs.map(g => glob.sync(g, {cwd: outdir}))
         // Reduce the array of arrays to a single array containing all the files that should be deleted
         .reduce((all, curr) => all.concat(curr), [])
)
```

Reply:
No, the logic uses function `matchFolderGlobs` in L72 to do folder check. Or just I did not understand the function.

Reply:
The logic is firstly globbing files/folders into a list of strings at line 53. The glob checks in line 72 are to ensure the folder we are going to remove is demanded by the user. That is:

- For a file, we only glob once (at line 53).
- For a folder, before removing it, we check if it matches the glob pattern for the second time. To ensure the glob pattern is **EXPLICITLY** specified by the user rather than `abc/**/`。

Reply:
The second check shall also be before the for loop, preferrably in a stream. Also, what if `abc/**/def/`?

Reply:
> Also, what if `abc/**/def/`?

In this implementation, we will remove it if it's empty.

Reply:
1. I am worried on `abc/**/def` case. If possible, none of `**` should be allowed.
2. The whole logic should be decoupled. First phase detect files and unlink them. Second phase detect folders (empty) and unlink them.

Reply:
> 1. I am worried on `abc/**/def` case. If possible, none of `**` should be allowed.

The user needs to add a slash `/` to the pattern explicitly as `abc/**/def/`. This would be intentional. I think it's fine. Also, we will prepend the patterns with `./**/` if set `latex.clean.subfolder.enabled` to `true`, it also contains globstar `**`.

> 2. The whole logic should be decoupled. First phase detect files and unlink them. Second phase detect folders (empty) and unlink them.

I'm fine with that.

Reply:
> The whole logic should be decoupled. First phase detect files and unlink them. Second phase detect folders (empty) and unlink them.

I found that if we decouple this, we will not need to sort the results by dictionary order any more.